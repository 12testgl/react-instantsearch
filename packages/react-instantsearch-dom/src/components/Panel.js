import React, { Component, createContext } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { createClassNames } from '../core/utils';

const cx = createClassNames('Panel');

export const {
  Consumer: PanelConsumer,
  Provider: PanelProvider,
} = createContext(function setCanRefine() {});

class Panel extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    header: PropTypes.node,
    footer: PropTypes.node,
  };

  static defaultProps = {
    className: '',
    header: null,
    footer: null,
  };

  state = {
    canRefine: true,
  };

  setCanRefine = nextCanRefine => {
    this.setState({ canRefine: nextCanRefine });
  };

  render() {
    const { children, className, header, footer } = this.props;
    const { canRefine } = this.state;

    return (
      <div
        className={classNames(cx('', !canRefine && '-noRefinement'), className)}
      >
        {header && <div className={cx('header')}>{header}</div>}

        <div className={cx('body')}>
          {React.Children.map(children, child => {
            if (!child.type || typeof child.type === 'string') return child;
            return React.cloneElement(child, {
              setCanRefine: this.setCanRefine,
            });
          })}
        </div>

        {footer && <div className={cx('footer')}>{footer}</div>}
      </div>
    );
  }
}

export default Panel;
