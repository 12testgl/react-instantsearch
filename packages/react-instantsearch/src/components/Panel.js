import React from 'react';
import PropTypes from 'prop-types';
import classNames from './classNames';

const Panel = ({ children, canRefine, cx, header, footer }) => (
  <div className={cx('', !canRefine && '-noRefinement')}>
    {header && <div className={`${cx('header')} ais-header`}>{header}</div>}

    <div className={`${cx('body')} ais-body`}>{children}</div>

    {footer && <div className={`${cx('footer')} ais-footer`}>{footer}</div>}
  </div>
);

Panel.propTypes = {
  children: PropTypes.node.isRequired,
  canRefine: PropTypes.bool,
  cx: PropTypes.func,
  header: PropTypes.node,
  footer: PropTypes.node,
};

Panel.defaultProps = {
  canRefine: true,
  cx: classNames('Panel'),
};

export default Panel;
