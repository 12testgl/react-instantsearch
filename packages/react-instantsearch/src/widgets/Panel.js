import Panel from '../components/Panel.js';

/**
 * The Panel widget wraps other widgets in a consistent panel design. This component wrap by default
 * all the built in widgets. It act as a skeleton for each widget by providing to them header, body and footer.
 * Since it wraps all the built in widgets by default you should rarely use it directly, but we still
 * provide it as a standaone widget to enable the possibility to wrap you custom widget with it.
 *
 * @name Panel
 * @kind widget
 * @propType {node} [header] - Adds a header to the widget.
 * @propType {node} [footer] - Adds a footer to the widget.
 * @themeKey ais-Panel - the root div of the Panel
 * @themeKey ais-Panel--noRefinement - the root div of the Panel without refinement
 * @themeKey ais-Panel-header - the header of the Panel (optional)
 * @themeKey ais-Panel-body - the body of the Panel
 * @themeKey ais-Panel-footer - the footer of the Panel (optional)
 * @example
 * import React from 'react';
 *
 * import { Panel, InstantSearch } from 'react-instantsearch/dom';
 * import { connectRefinementList } from 'react-instantsearch/connectors';
 *
 * const CustomRefinementList = connectRefinementList(({ header }) => (
 *   <Panel header={header}>
 *     <div>My custom refinement list</div>
 *   </Panel>
 * ));
 *
 * export default function App() {
 *   return (
 *     <InstantSearch
 *       appId="latency"
 *       apiKey="6be0576ff61c053d5f9a3225e2a90f76"
 *       indexName="ikea"
 *     >
 *      <CustomRefinementList
 *        attributeName="category"
 *        header="Category"
 *      />
 *     </InstantSearch>
 *   );
 * }
 */
export default Panel;
