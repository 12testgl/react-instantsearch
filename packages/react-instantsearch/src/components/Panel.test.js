import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Panel from './Panel';

Enzyme.configure({ adapter: new Adapter() });

describe('Panel', () => {
  it('expect to render', () => {
    const wrapper = shallow(
      <Panel>
        <div>Hello content</div>
      </Panel>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('expect to render with custom className', () => {
    const props = {
      className: 'ais-Panel-Breadcrumb',
    };

    const wrapper = shallow(
      <Panel {...props}>
        <div>Hello content</div>
      </Panel>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('expect to render without refinement', () => {
    const props = {
      canRefine: false,
    };

    const wrapper = shallow(
      <Panel {...props}>
        <div>Hello content</div>
      </Panel>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('expect to render with header', () => {
    const props = {
      header: <p>Header</p>,
    };

    const wrapper = shallow(
      <Panel {...props}>
        <div>Hello content</div>
      </Panel>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('expect to render with footer', () => {
    const props = {
      footer: <p>Footer</p>,
    };

    const wrapper = shallow(
      <Panel {...props}>
        <div>Hello content</div>
      </Panel>
    );

    expect(wrapper).toMatchSnapshot();
  });
});
