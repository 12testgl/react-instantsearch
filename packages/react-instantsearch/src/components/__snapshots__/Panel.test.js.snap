// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel expect to render 1`] = `
<div
  className="ais-Panel"
>
  <div
    className="ais-Panel-body ais-body"
  >
    <div>
      Hello content
    </div>
  </div>
</div>
`;

exports[`Panel expect to render with footer 1`] = `
<div
  className="ais-Panel"
>
  <div
    className="ais-Panel-body ais-body"
  >
    <div>
      Hello content
    </div>
  </div>
  <div
    className="ais-Panel-footer ais-footer"
  >
    <p>
      Footer
    </p>
  </div>
</div>
`;

exports[`Panel expect to render with header 1`] = `
<div
  className="ais-Panel"
>
  <div
    className="ais-Panel-header ais-header"
  >
    <p>
      Header
    </p>
  </div>
  <div
    className="ais-Panel-body ais-body"
  >
    <div>
      Hello content
    </div>
  </div>
</div>
`;

exports[`Panel expect to render without refinement 1`] = `
<div
  className="ais-Panel ais-Panel--noRefinement"
>
  <div
    className="ais-Panel-body ais-body"
  >
    <div>
      Hello content
    </div>
  </div>
</div>
`;
